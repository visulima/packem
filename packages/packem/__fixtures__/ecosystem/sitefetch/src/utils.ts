import micromatch from "micromatch"

// xK or xM
export function formatNumber(num: number): string {
  return num > 1000000
    ? `${(num / 1000000).toFixed(1)}M`
    : num > 1000
    ? `${(num / 1000).toFixed(1)}K`
    : num.toString()
}

export function matchPath(path: string, pattern: string | string[]): boolean {
  return micromatch.isMatch(path, pattern)
}

export function ensureArray<T>(input: T | T[]): T[] {
  return Array.isArray(input) ? input : [input]
}
