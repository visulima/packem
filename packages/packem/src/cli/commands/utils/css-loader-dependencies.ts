const cssLoaderDependencies: Record<string, string[]> = {
    less: ["less"],
    lightningcss: ["lightningcss"],
    "node-sass": ["node-sass"],
    postcss: [
        "postcss",
        "postcss-load-config",
        "postcss-modules-extract-imports",
        "postcss-modules-local-by-default",
        "postcss-modules-scope",
        "postcss-modules-values",
        "postcss-value-parser",
        "icss-utils",
    ],
    sass: ["sass"],
    "sass-embedded": ["sass-embedded"],
    stylus: ["stylus"],
};

export default cssLoaderDependencies;
