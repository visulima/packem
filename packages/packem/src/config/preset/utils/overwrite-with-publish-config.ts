import type { NormalizedPackageJson } from "@visulima/package";

import type { BuildConfig } from "../../../types";

// eslint-disable-next-line sonarjs/cognitive-complexity
const overwriteWithPublishConfig = (packageJson: NormalizedPackageJson, declaration: BuildConfig["declaration"]): NormalizedPackageJson => {
    const { publishConfig } = packageJson;

    if (publishConfig) {
        if (publishConfig.bin && (typeof publishConfig.bin === "object" || typeof publishConfig.bin === "string")) {
            // eslint-disable-next-line no-param-reassign
            packageJson.bin = publishConfig.bin as NormalizedPackageJson["bin"];
        }

        if (publishConfig.type && typeof publishConfig.type === "string" && publishConfig.type !== "") {
            // eslint-disable-next-line no-param-reassign
            packageJson.type = publishConfig.type as NormalizedPackageJson["type"];
        }

        if (publishConfig.main && typeof publishConfig.main === "string" && publishConfig.main !== "") {
            // eslint-disable-next-line no-param-reassign
            packageJson.main = publishConfig.main as NormalizedPackageJson["main"];
        }

        if (publishConfig.module && typeof publishConfig.module === "string" && publishConfig.module !== "") {
            // eslint-disable-next-line no-param-reassign
            packageJson.module = publishConfig.module as NormalizedPackageJson["module"];
        }

        if (declaration === undefined && publishConfig.types && typeof publishConfig.types === "string" && publishConfig.types !== "") {
            // eslint-disable-next-line no-param-reassign
            packageJson.types = publishConfig.types as NormalizedPackageJson["types"];
        } else if (declaration === undefined && publishConfig.typings && typeof publishConfig.typings === "string" && publishConfig.typings !== "") {
            // eslint-disable-next-line no-param-reassign
            packageJson.typings = publishConfig.typings as NormalizedPackageJson["typings"];
        }

        if (publishConfig.exports && typeof publishConfig.exports === "object") {
            // eslint-disable-next-line no-param-reassign
            packageJson.exports = publishConfig.exports as NormalizedPackageJson["exports"];
        }
    }

    return packageJson;
};

export default overwriteWithPublishConfig;
