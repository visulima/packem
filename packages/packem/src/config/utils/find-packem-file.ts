import { isAccessible } from "@visulima/fs";
import { join } from "@visulima/path";

const findPackemFile = async (rootDirectory: string, configPath = ""): Promise<string> => {
    let packemConfigFilePath = configPath;

    if (!packemConfigFilePath) {
        const packemConfigFiles = ["packem.config.js", "packem.config.mjs", "packem.config.cjs", "packem.config.ts", "packem.config.cts", "packem.config.mts"];

        for (const file of packemConfigFiles) {
            // eslint-disable-next-line no-await-in-loop
            if (await isAccessible(join(rootDirectory, file))) {
                packemConfigFilePath = "./" + file;
                break;
            }
        }
    }

    if (!/\.(?:js|mjs|cjs|ts|cts|mts)$/.test(packemConfigFilePath)) {
        throw new Error("Invalid packem config file extension. Only .js, .mjs, .cjs, .ts, .cts and .mts extensions are allowed.");
    }

    return packemConfigFilePath;
};

export default findPackemFile;
