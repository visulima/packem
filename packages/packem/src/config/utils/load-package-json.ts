import { isAccessibleSync } from "@visulima/fs";
import type { NormalizedPackageJson } from "@visulima/package";
import { parsePackageJson } from "@visulima/package/package-json";
import { join } from "@visulima/path";

const loadPackageJson = (
    rootDirectory: string,
): {
    packageJson: NormalizedPackageJson;
    packageJsonPath: string;
} => {
    const packageJsonPath = join(rootDirectory, "package.json");

    if (!isAccessibleSync(packageJsonPath)) {
        throw new Error(`package.json not found at ${packageJsonPath}`);
    }

    const packageJson = parsePackageJson(packageJsonPath);

    return {
        packageJson,
        packageJsonPath,
    };
};

export default loadPackageJson;
