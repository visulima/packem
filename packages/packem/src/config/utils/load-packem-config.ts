import type { Jiti } from "jiti";

import type { BuildConfig, BuildConfigFunction, Environment, Mode } from "../../types";
import findPackemFile from "./find-packem-file";

const loadPackemConfig = async (
    jiti: Jiti,
    rootDirectory: string,
    environment: Environment,
    mode: Mode,
    configPath?: string,
): Promise<{
    config: BuildConfig;
    path: string;
}> => {
    const packemConfigFilePath = await findPackemFile(rootDirectory, configPath);

    let buildConfig = (await jiti.import(packemConfigFilePath, { default: true, try: true }) || {}) as BuildConfig | BuildConfigFunction;

    if (typeof buildConfig === "function") {
        buildConfig = await buildConfig(environment, mode);
    }

    return {
        config: buildConfig,
        path: packemConfigFilePath,
    };
};

export default loadPackemConfig;
