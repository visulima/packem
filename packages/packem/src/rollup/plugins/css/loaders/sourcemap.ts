export { default } from "@visulima/rollup-css-plugin/loader/sourcemap";
