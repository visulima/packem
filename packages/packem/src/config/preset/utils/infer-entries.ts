import { isAccessibleSync } from "@visulima/fs";
import type { PackageJson } from "@visulima/package";
import { extname, resolve, toNamespacedPath } from "@visulima/path";

import { DEFAULT_EXTENSIONS, DEVELOPMENT_ENV, ENDING_REGEX, PRODUCTION_ENV, RUNTIME_EXPORT_CONVENTIONS, SPECIAL_EXPORT_CONVENTIONS } from "../../../constants";
import type { BuildContext, BuildEntry, Environment, InferEntriesResult, Runtime } from "../../../types";
import type { OutputDescriptor } from "../../../utils/extract-export-filenames";
import { extractExportFilenames } from "../../../utils/extract-export-filenames";
import { inferExportTypeFromFileName } from "../../../utils/infer-export-type";

const getEnvironment = (output: OutputDescriptor, environment: Environment): Environment => {
    if (output.key === "exports" && output.subKey === PRODUCTION_ENV) {
        return PRODUCTION_ENV;
    }

    if (output.key === "exports" && output.subKey === DEVELOPMENT_ENV) {
        return DEVELOPMENT_ENV;
    }

    return environment;
};

const createOrUpdateEntry = (
    entries: BuildEntry[],
    input: string,
    isDirectory: boolean,
    outputSlug: string,
    output: OutputDescriptor,
    context: BuildContext,
    isGlob: boolean,
    // eslint-disable-next-line sonarjs/cognitive-complexity
): void => {
    const entryEnvironment = getEnvironment(output, context.environment);

    let runtime: Runtime = context.options.runtime as Runtime;

    for (const runtimeExportConvention of RUNTIME_EXPORT_CONVENTIONS) {
        if (output.file.includes("." + runtimeExportConvention + ".") || output.subKey === runtimeExportConvention) {
            runtime = runtimeExportConvention as Runtime;

            break;
        }
    }

    let entry: BuildEntry | undefined = entries.find((index) => index.input === input && index.environment === entryEnvironment && index.runtime === runtime);

    if (entry === undefined) {
        entry = entries[
            entries.push({ environment: entryEnvironment, exportKey: new Set([output.exportKey].filter(Boolean)), input, runtime }) - 1
        ] as BuildEntry;
    } else if (entry.exportKey && output.exportKey) {
        entry.exportKey.add(output.exportKey);
    }

    if (isGlob) {
        entry.isGlob = true;
    }

    if (isDirectory) {
        entry.outDir = outputSlug;
    }

    if (output.isExecutable) {
        entry.executable = true;

        entry.declaration = false;

        if (output.type === "cjs") {
            entry.cjs = true;
        } else if (output.type === "esm") {
            entry.esm = true;
        }
    } else {
        if (/\.d\.[mc]?ts$/.test(output.file) && context.options.declaration !== false) {
            entry.declaration = context.options.declaration;
        }

        if (output.type === "cjs") {
            entry.cjs = true;
        } else if (output.type === "esm") {
            entry.esm = true;
        }
    }

    // eslint-disable-next-line @rushstack/security/no-unsafe-regexp,security/detect-non-literal-regexp
    const aliasName = output.file.replace(extname(output.file), "").replace(new RegExp(`^./${context.options.outDir.replace(/^\.\//, "")}/`), "");

    if (SPECIAL_EXPORT_CONVENTIONS.has(output.subKey as string) && !input.includes(aliasName)) {
        entry.fileAlias = aliasName;
    }
};

let privateSubfolderWarningShown = false;

const validateIfTypescriptIsInstalled = (context: BuildContext): void => {
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    if (context.pkg?.dependencies?.typescript === undefined && context.pkg?.devDependencies?.typescript === undefined) {
        // @TODO: Add command to install typescript
        throw new Error("You tried to use a `.ts`, `.cts` or `.mts` file but `typescript` was not found in your package.json.");
    }
};

/**
 * Infer entries from package files.
 *
 * @param {PackageJson} packageJson
 * @param {string[]} sourceFiles A list of source files to use for inferring entries.
 * @param {BuildContext} context
 * @returns {InferEntriesResult}
 */
const inferEntries = (
    packageJson: PackageJson,
    sourceFiles: string[],
    context: BuildContext,
    // eslint-disable-next-line sonarjs/cognitive-complexity
): InferEntriesResult => {
    const warnings: string[] = [];

    // Sort files so least-nested files are first
    sourceFiles.sort((a, b) => a.split("/").length - b.split("/").length);

    const packageType = packageJson.type === "module" ? "esm" : "cjs";

    if (packageType === "esm") {
        context.options.emitESM = true;
    }

    if (packageType === "cjs") {
        context.options.emitCJS = true;
    }

    // Come up with a list of all output files & their formats
    const outputs = extractExportFilenames(packageJson.exports, packageType, context.options.declaration);

    if (packageJson.bin) {
        const binaries = (typeof packageJson.bin === "string" ? [packageJson.bin] : Object.values(packageJson.bin)).filter(Boolean);

        for (const file of binaries) {
            const inferredType = inferExportTypeFromFileName(file);

            if (inferredType && inferredType !== packageType) {
                throw new Error(`Exported file "${file}" has an extension that does not match the package.json type "${packageJson.type ?? "commonjs"}".`);
            }

            outputs.push({ file: file as string, isExecutable: true, key: "bin", type: inferredType ?? packageType });
        }
    }

    if (packageJson.main) {
        outputs.push({
            file: packageJson.main,
            key: "main",
            type: inferExportTypeFromFileName(packageJson.main) ?? packageType,
        });
    }

    // Defacto module entry-point for bundlers (not Node.js)
    // https://github.com/dherman/defense-of-dot-js/blob/master/proposal.md
    if (packageJson.module) {
        outputs.push({ file: packageJson.module, key: "module", type: "esm" });
    }

    // Entry point for TypeScript
    if (packageJson.types || packageJson.typings) {
        validateIfTypescriptIsInstalled(context);

        if (context.options.declaration === undefined) {
            context.options.declaration = "compatible";
        }

        outputs.push({ file: (packageJson.types ?? packageJson.typings) as string, key: "types" });
    }

    // Infer entries from package files
    const entries: BuildEntry[] = [];

    for (const output of outputs) {
        const outputExtension = extname(output.file);

        // Only javascript files are supported
        if (outputExtension !== "" && !DEFAULT_EXTENSIONS.includes(outputExtension)) {
            // eslint-disable-next-line no-continue
            continue;
        }

        if (context.options.declaration === undefined && (output.key === "types" || output.subKey === "types")) {
            context.options.declaration = output.file.includes(".d.ts") ? "compatible" : true;
        }

        if (context.options.emitCJS === undefined && output.type === "cjs") {
            context.options.emitCJS = true;
        }

        if (context.options.emitESM === undefined && output.type === "esm") {
            context.options.emitESM = true;
        }

        // Supported output file extensions are `.d.ts`, `.cjs` and `.mjs`
        // But we support any file extension here in case user has extended rollup options
        const outputSlug = output.file.replace(/(?:\*[^/\\]|\.d\.[mc]?ts|\.\w+)$/, "");
        const isDirectory = outputSlug.endsWith("/");

        // Skip top level directory
        if (isDirectory && ["./", "/"].includes(outputSlug)) {
            // eslint-disable-next-line no-continue
            continue;
        }

        // eslint-disable-next-line @rushstack/security/no-unsafe-regexp,security/detect-non-literal-regexp
        const sourceSlug = outputSlug.replace(new RegExp("(./)?" + context.options.outDir), context.options.sourceDir).replace("./", "");

        const beforeSourceRegex = "(?<=/|$)";
        const fileExtensionRegex = isDirectory ? "" : "(\\.d\\.[cm]?ts|(\\.[cm]?[tj]sx?))$";

        // @see https://nodejs.org/docs/latest-v16.x/api/packages.html#subpath-patterns
        if (output.file.includes("/*") && output.key === "exports") {
            if (!privateSubfolderWarningShown) {
                context.logger.debug("Private subfolders are not supported, if you need this feature please open an issue on GitHub.");

                privateSubfolderWarningShown = true;
            }

            const inputs: string[] = [];

            // eslint-disable-next-line @rushstack/security/no-unsafe-regexp,security/detect-non-literal-regexp
            const SOURCE_RE = new RegExp(beforeSourceRegex + sourceSlug.replace("*", "(.*)") + fileExtensionRegex);
            // eslint-disable-next-line @rushstack/security/no-unsafe-regexp,security/detect-non-literal-regexp
            const SPECIAL_SOURCE_RE = new RegExp(beforeSourceRegex + sourceSlug.replace(/(.*)\.[^.]*$/, "$1").replace("*", "(.*)") + fileExtensionRegex);

            for (const source of sourceFiles) {
                if (SOURCE_RE.test(source) || (SPECIAL_EXPORT_CONVENTIONS.has(output.subKey as string) && SPECIAL_SOURCE_RE.test(source))) {
                    inputs.push(source);
                }
            }

            if (inputs.length === 0) {
                warnings.push(`Could not find entrypoint for \`${output.file}\``);

                // eslint-disable-next-line no-continue
                continue;
            }

            for (const input of inputs) {
                createOrUpdateEntry(entries, input, isDirectory, outputSlug, output, context, true);
            }

            // eslint-disable-next-line no-continue
            continue;
        }

        // eslint-disable-next-line @rushstack/security/no-unsafe-regexp,security/detect-non-literal-regexp
        const SOURCE_RE = new RegExp(beforeSourceRegex + sourceSlug + fileExtensionRegex);

        let input = sourceFiles.find((index) => SOURCE_RE.test(index));

        if (SPECIAL_EXPORT_CONVENTIONS.has(output.subKey as string) && input === undefined) {
            // eslint-disable-next-line @rushstack/security/no-unsafe-regexp,security/detect-non-literal-regexp
            const SPECIAL_SOURCE_RE = new RegExp(beforeSourceRegex + sourceSlug.replace(/(.*)\.[^.]*$/, "$1") + fileExtensionRegex);

            input = sourceFiles.find((index) => SPECIAL_SOURCE_RE.test(index));
        }

        if (input === undefined) {
            if (!isAccessibleSync(resolve(context.options.rootDir, output.file))) {
                warnings.push(`Could not find entrypoint for \`${output.file}\``);
            }

            // eslint-disable-next-line no-continue
            continue;
        }

        if ((input.endsWith(".ts") || input.endsWith(".cts") || input.endsWith(".mts")) && isAccessibleSync(input)) {
            validateIfTypescriptIsInstalled(context);
        }

        const inputWithoutExtension = toNamespacedPath(input.replace(ENDING_REGEX, ""));

        if (isAccessibleSync(inputWithoutExtension + ".cts") && isAccessibleSync(inputWithoutExtension + ".mts")) {
            createOrUpdateEntry(entries, inputWithoutExtension + ".cts", isDirectory, outputSlug, { ...output, type: "cjs" }, context, false);
            createOrUpdateEntry(entries, inputWithoutExtension + ".mts", isDirectory, outputSlug, { ...output, type: "esm" }, context, false);
        } else {
            createOrUpdateEntry(entries, input, isDirectory, outputSlug, output, context, false);
        }
    }

    return { entries, warnings };
};

export default inferEntries;
