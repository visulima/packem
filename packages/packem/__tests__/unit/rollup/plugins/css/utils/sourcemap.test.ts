import { join } from "@visulima/path";
import { describe, expect, it } from "vitest";

import { getMap, mm, stripMap } from "../../../../../../src/rollup/plugins/css/utils/sourcemap";

const fixturePath = join(__dirname, "../../../../../../", "__fixtures__", "css");

describe("sourcemap", () => {
    it("inline map", async () => {
        expect.assertions(3);

        const code
            = ".foo {color: red;background-image: url(\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkBAMAAACCzIhnAAAAG1BMVEXMzMyWlpacnJy+vr6jo6PFxcW3t7eqqqqxsbHbm8QuAAAACXBIWXMAAA7EAAAOxAGVKw4bAAAAiklEQVRYhe3QMQ6EIBAF0C+GSInF9mYTs+1ewRsQbmBlayysKefYO2asXbbYxvxHQj6ECQMAEREREf2NQ/fCtp5Zky6vtRMkSJEzhyISynWJnzH6Z8oQlzS7lEc/fLmmQUSvc16OrCPqRl1JePxQYo1ZSWVj9nxrrOb5esw+eXdvzTWfTERERHRXH4tWFZGswQ2yAAAAAElFTkSuQmCC\");}";

        await expect(getMap(code)).resolves.toBeUndefined();

        const withBlock = `${code}/*# sourceMappingURL=data:application/json;base64,e1RISVM6SVNBU09VUkNFTUFQU0lNVUxBVElPTn0= */`;

        await expect(getMap(withBlock)).resolves.toBe("{THIS:ISASOURCEMAPSIMULATION}");

        const withInline = `${code}//# sourceMappingURL=data:application/json;base64,e1RISVM6SVNBU09VUkNFTUFQU0lNVUxBVElPTn0=`;

        await expect(getMap(withInline)).resolves.toBe("{THIS:ISASOURCEMAPSIMULATION}");
    });

    it("file map", async () => {
        expect.assertions(3);

        const code = ".foo {color: red;}/*# sourceMappingURL=fixture.css.map */";

        await expect(getMap(code)).rejects.toThrowErrorMatchingSnapshot();
        await expect(getMap(code, "this/is/nonexistant/path.css")).resolves.toBeUndefined();
        await expect(getMap(code, join(fixturePath, "utils/pointless.css"))).resolves.toBe("{THIS:ISASOURCEMAPSIMULATION}");
    });

    it("strip map", () => {
        expect.assertions(2);

        const code = ".foo {color: red;}";
        const withBlock = `${code}/*# sourceMappingURL=fixture.css.map */`;

        expect(stripMap(withBlock)).toBe(".foo {color: red;}");

        const withInline = `${code}//# sourceMappingURL=fixture.css.map`;

        expect(stripMap(withInline)).toBe(".foo {color: red;}");
    });

    it("map modifier", () => {
        expect.assertions(2);

        const map = JSON.stringify({ sources: ["../a/b/../foo/bar.css", "../b/a/../bar/foo.css"] });
        const relativeSource = JSON.stringify(mm(map).relative().toObject()?.sources);

        expect(relativeSource).toBe(JSON.stringify(["../a/foo/bar.css", "../b/bar/foo.css"]));
        expect(mm("thisisnotjson").toString()).toBeUndefined();
    });
});
