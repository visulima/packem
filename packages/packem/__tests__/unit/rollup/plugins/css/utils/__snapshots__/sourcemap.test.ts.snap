// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`css Sourcemap Utilities > should correctly handle file-based sourcemaps 1`] = `[Error: Extracted map detected, but no ID is provided]`;

exports[`sourcemap > file map 1`] = `[Error: Extracted map detected, but no ID is provided]`;
